ssl/t1_lib.c

```diff
@@ -1260,44 +1282,31 @@ static int tls_decrypt_ticket(SSL *s, const unsigned char *etick,
 int tls12_get_sigandhash(SSL *s, WPACKET *pkt, const EVP_PKEY *pk,
                          const EVP_MD *md, int *ispss)
 {
-    int md_id, sig_id, tmpispss = 0;
+    int md_id, sig_id;
     size_t i;
-    const SIGALG_LOOKUP *curr;
 
     if (md == NULL)
         return 0;
     md_id = EVP_MD_type(md);
     sig_id = EVP_PKEY_id(pk);
     if (md_id == NID_undef)
         return 0;
+    /* For TLS 1.3 only allow RSA-PSS */
+    if (SSL_IS_TLS13(s) && sig_id == EVP_PKEY_RSA)
+        sig_id = EVP_PKEY_RSA_PSS;
 
-    for (i = 0, curr = sigalg_lookup_tbl; i < OSSL_NELEM(sigalg_lookup_tbl);
-         i++, curr++) {
-        if (curr->hash == md_id && curr->sig == sig_id) {
-            if (sig_id == EVP_PKEY_RSA) {
-                tmpispss = SIGID_IS_PSS(curr->sigalg);
-                if (!SSL_IS_TLS13(s) && tmpispss) {
-                    size_t j;
-
-                    /*
-                     * Check peer actually sent a PSS sig id - it could have
-                     * been a PKCS1 sig id instead.
-                     */
-                    for (j = 0; j < s->cert->shared_sigalgslen; j++)
-                        if (s->cert->shared_sigalgs[j].rsigalg == curr->sigalg)
-                            break;
-
-                    if (j == s->cert->shared_sigalgslen)
-                        continue;
-                }
-            }
+    for (i = 0; i < s->cert->shared_sigalgslen; i++) {
+        const SIGALG_LOOKUP *curr = s->cert->shared_sigalgs[i];
+
+        /* If key type is RSA also match PSS signature type */
+        if ((sig_id == EVP_PKEY_RSA && curr->sig == EVP_PKEY_RSA_PSS)
+            || (curr->hash == md_id && curr->sig == sig_id)) {
```

Comment:
Doesn't that mean we don't match on the hash if `sig_id == EVP_PKEY_RSA && curr->sig == EVP_PKEY_RSA_PSS`